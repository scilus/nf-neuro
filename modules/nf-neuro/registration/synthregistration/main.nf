process REGISTRATION_SYNTHREGISTRATION {
    tag "$meta.id"
    label 'process_high'

    container "freesurfer/synthmorph:4"
    containerOptions {
        (workflow.containerEngine == 'docker') ? '--entrypoint "" --env PYTHONPATH="/freesurfer/env/lib/python3.11/site-packages"' : "--env PYTHONPATH='/freesurfer/env/lib/python3.11/site-packages'"
    }

    input:
    tuple val(meta), path(moving), path(fixed)

    output:
    tuple val(meta), path("*__output_warped.nii.gz")        , emit: warped_image
    tuple val(meta), path("*__deform_warp.nii.gz")          , emit: warp
    tuple val(meta), path("*__affine_warp.lta")             , emit: affine
    path "versions.yml"                                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def affine = task.ext.affine ? "-m " + task.ext.affine : "-m affine"
    def warp = task.ext.warp ? "-m " + task.ext.warp : "-m deform"
    def header = task.ext.header ? "-H" : ""
    def gpu = task.ext.gpu ? "-g" : ""
    def lambda = task.ext.lambda ? "-r " + task.ext.lambda : ""
    def steps = task.ext.steps ? "-n " + task.ext.steps : ""
    def extent = task.ext.extent ? "-e " + task.ext.extent : ""
    def weight = task.ext.weight ? "-w " + task.ext.weight : ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1
    mri_synthmorph -j $task.cpus ${affine} -t ${prefix}__affine_warp.lta $moving $fixed
    mri_synthmorph -j $task.cpus ${warp} ${gpu} ${lambda} ${steps} ${extent} ${weight} -i ${prefix}__affine_warp.lta  -t ${prefix}__deform_warp.nii.gz -o ${prefix}__output_warped.nii.gz $moving $fixed

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        synthmoprh: 4
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mri_synthmorph -h

    touch ${prefix}__output_warped.nii.gz
    touch ${prefix}__deform_warp.nii.gz
    touch ${prefix}__affine_warp.lta

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        synthmoprh: 4
    END_VERSIONS
    """
}
