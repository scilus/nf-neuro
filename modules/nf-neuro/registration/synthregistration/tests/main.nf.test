nextflow_process {

    name "Test Process REGISTRATION_SYNTHREGISTRATION"
    script "../main.nf"
    process "REGISTRATION_SYNTHREGISTRATION"

    tag "modules"
    tag "modules_nfcore"
    tag "registration"
    tag "registration/synthregistration"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    // ** Assertion only uses file name since there is a discrepancy when tests are ran remotly. ** //
    // ** Synthmorph is an AI model and freesurfer does not control for randomness in computation yet, both in terms of seeding and operation ordering.** //

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "processing.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("registration - synthregistration") {
        config "./nextflow.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                    [ id:'test', single_end:false ],
                    file("\${test_data_directory}/mni_masked_2x2x2.nii.gz"),
                    file("\${test_data_directory}/b0_mean.nii.gz")
                ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.warped_image.get(0).get(1)).name,
                    file(process.out.warp.get(0).get(1)).name,
                    file(process.out.affine.get(0).get(1)).name,
                    process.out.versions
                ).match() }
            )
        }
    }
}
