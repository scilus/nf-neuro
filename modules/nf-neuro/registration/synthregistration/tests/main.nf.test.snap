{
    "registration - synthregistration": {
        "content": [
            "test__output_warped.nii.gz",
            "test__deform_warp.nii.gz",
            "test__affine_warp.lta",
            [
                "versions.yml:md5,6f946408922f25f1a8182d35615b6d7b"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-12-18T13:10:28.733204"
    }
}
